packages/@sanity/schema/src/sanity/extractSchema.ts

```diff
@@ -0,0 +1,669 @@
+import type {
+  SchemaTypeDefinition,
+  DocumentDefinition,
+  ObjectDefinition,
+  ArrayDefinition,
+  FieldDefinition,
+  BlockDefinition,
+  ArrayOfType,
+  ReferenceDefinition,
+  ImageDefinition,
+  FileDefinition,
+  CrossDatasetReferenceDefinition,
+} from '@sanity/types'
+
+// TODO: Move these types to groq-js
+interface Document {
+  type: 'document'
+  name: string
+  fields: ObjectKeyValueField<Field>[]
+}
+
+interface TypeDecleration {
+  type: 'type'
+  name: string
+  value: Field
+}
+
+type SchemaField = Document | TypeDecleration
+type Schema = SchemaField[]
+
+interface StringField {
+  type: 'string'
+  value?: string
+}
+interface NumberField {
+  type: 'number'
+  value?: number
+}
+interface BooleanField {
+  type: 'boolean'
+  value?: boolean
+}
+interface NullField {
+  type: 'null'
+}
+
+interface ObjectField<T extends Field = Field> {
+  type: 'object'
+  fields: ObjectKeyValueField<T>[]
+}
+
+interface ObjectKeyValueField<T extends Field = Field> {
+  type: 'objectKeyValue'
+  key: string
+  value: T
+}
+
+interface ReferenceField {
+  type: 'reference'
+  to: string
+  childOf?: ReferenceField
+  resolved?: boolean
+}
+
+interface UnionField<T extends Field = Field> {
+  type: 'union'
+  of: T[]
+}
+
+interface ArrayField<T extends Field = Field> {
+  type: 'array'
+  of: T
+}
+interface ConcatenationField {
```

Comment:
What is a concatenation field?

Reply:
concatenation is a set of fields concatenated together :D 

ie. the groq query `"foo" + 3 + name` would be:
```
{
type: "concatenation",
fields: [
  { type: "string", value: "foo" },
  { type: "number", value: 3 },
  { type: "string" }, # or number/boolean depending on the schema
],
}